### PR TITLE
Simplex DC meshing

### PR DESCRIPTION
**Pull request by Yitzi2**

### CODE REVIEW

**Comment by cupe-ntop** [comment]
Is this a libfive update from upstream or is this our contribution? I scrolled through the changes to give it a glance, but it's impossible to really understand what's going on in 9K lines of diff without spending at least a day on this. Not sure if I should do that, especially since the change looks like it doesn't touch existing functionality for the most part.

**Comment by cupe-ntop** [comment]
I have added some repro files (.ntop files, screenshots, and screengrab videos) here: https://drive.google.com/drive/folders/1KAd6haHw9O5VaK6PftBwJWrnP4B1Gxtj?usp=sharing

File: libfive/src/tree/opcode.cpp [hunk 1]
```diff
@@ -226,4 +226,4 @@ bool Opcode::isCommutative(Opcode op)
     return false;
 }
 
-}   // namespace libfive
+}   // namespace libfive
```
Comment (cupe-ntop) [inline libfive/src/tree/opcode.cpp:229]: probably not an intended change

**Comment by cupe-ntop** [comment]
I find this heavily-templated code very hard to follow and would have trouble maintaining it. Not sure if that could be changed. Also it may be helpful to link the publications that this approach builds upon in some code comment somewhere?

**Comment by Yitzi2** [comment]
> Is this a libfive update from upstream or is this our contribution? I scrolled through the changes to give it a glance, but it's impossible to really understand what's going on in 9K lines of diff without spending at least a day on this. Not sure if I should do that, especially since the change looks like it doesn't touch existing functionality for the most part.

This is ours.

File: libfive/src/render/brep/mesh.cpp [hunk 2]
```diff
@@ -116,6 +126,115 @@ std::unique_ptr<Mesh> Mesh::render(
...
+        }
+
+        auto intersectionVerts = intersectionMap->buildVecs();
+
+        for (auto& a : intersectionVerts) {
+            for (auto& b : intersectionVerts) {
+                if (a == b && &a != &b) {
+                    throw std::pair(a, b);
+                }
+            }
```
[the remark this answers is not in the document]
Comment (Yitzi2) [reply libfive/src/render/brep/mesh.cpp:204]: This was test code, and keeping it in was an accident.  It is removed in my latest (local) branch.

File: libfive/src/render/brep/mesh.cpp [hunk 3]
```diff
@@ -116,6 +126,115 @@ std::unique_ptr<Mesh> Mesh::render(
                 });
         t.reset(settings);
     }
+    else if (settings.alg == SIMPLEX_DC) {
+        if (settings.progress_handler) {
+            // Pool::build, Edges, SubConsolidator Collapser 2 and 3, 
+            // Intersecter, Vertexer, Mesher, t.reset
+            settings.progress_handler->start({ 1, 1, 1, 1, 1, 1, 1, 1, 1 });
+        }
+        auto t = SimplexDCWorkerPool<3>::build(es, r, settings);
```
Comment (cupe-ntop) [inline libfive/src/render/brep/mesh.cpp:135]: I cancelled the task pretty quickly after starting it, and somehow one thread seemed to get stuck in ::build, inside ~Root() -> reset(). The task seems cancelled (progressbar is gone), but ntop behaves pretty sluggish. upon attaching the debugger, I'm now finding a single thread waiting (seemingly forever) for some futures, see screenshot below.
devenv_2020-05-27_20-02-49
